proposals/SelectorResourceBudgets.md

```diff
@@ -0,0 +1,207 @@
+# Selectors have sufficiently predictable resource budgets to be used in low-trust environments
+
+Authors: @warpfork
+
+Initial PR: https://github.com/protocol/web3-dev-team/pull/27
+
+
+Purpose & impact
+----------------
+
+#### Background & intent
+_Describe the desired state of the world after this project? Why does that matter?_
+
+The status quo is: we have Selectors, and they can be used to describe walks of graphs of data.
+(They're sorta like regexps for DAGs, if that's a useful comparison for you.)
+We want to expose these
+
+The problem is: if a service wants to accept Selectors which are user-specified,
```

Comment:
For go-ipfs: right, I could care less if someone uses a local API to hose _themselves_.  But people seem to want to expose these things publicly.  For example, https://github.com/protocol/web3-dev-team/pull/1 seems to imply we're going to have APIs oriented around Selector queries, and does not say much about not letting these be exposed remotely.  This is representative of most conversations I've ever overheard about Selectors and what people want to do with them.  (So, if nothing else, this proposal needed to be made to track the situation!)

For filecoin: I taaaag.... @magik6k ?  (I have repeatedly heard this is wanted, that's the depth limit of my knowing.)

In general: it seems like it's almost a law of human nature that people want to ask arbitrarily complex questions without concern for the costs on the answerer :laughing: / :cry:  The Selectors system seems to be no exception, heh.

Reply:
Big use-case I can't believe I forgot in the earlier comment: *graphsync*.

We seem to talk about the intention to use graphsync between untrusted peers who might be exchanging data without a fee mechanism.  If that's true, then it will be important for such peers to have at least some cost estimation mechanism and cutoff options.